compiler/optimizer/abstractinterpreter/OMRIDTBuilder.cpp

```diff
@@ -0,0 +1,324 @@
+
+   // eliminate call targets that are not inlinable according to the policy
+   // thus they won't be added to IDT
+   getInliner()->applyPolicyToTargets(callStack, callSite);
+
+   if (callSite->numTargets() == 0)
+      {
+      if (traceBIIDTGen)
+         traceMsg(comp(), "Do not have a call target. Don't add\n");
+      return;
+      }
+
+   for (int32_t i = 0 ; i < callSite->numTargets(); i++)
+      {
+      TR_CallTarget* callTarget = callSite->getTarget(i);
+
+      int32_t remainingBudget = parent->getBudget() - callTarget->_calleeMethod->maxBytecodeIndex();
+
+      if (remainingBudget < 0 ) // no budget remains
+         {
+         if (traceBIIDTGen)
+            traceMsg(comp(), "No budget left. Don't add\n");
+         continue;
+         }
+
+      bool isRecursiveCall = callStack->isAnywhereOnTheStack(callTarget->_calleeMethod, 1);
+
+      if (isRecursiveCall) //Stop for recursive call
+         {
+         if (traceBIIDTGen)
+            traceMsg(comp(), "Recursive call. Don't add\n");
+         continue;
+         }
+
+#ifdef J9_PROJECT_SPECIFIC
+      // hot and scorching bodies should never be inlined to warm or cooler bodies
+      if (!callTarget->_calleeMethod->isInterpreted())
+         {
+         TR_PersistentJittedBodyInfo * bodyInfo = ((TR_ResolvedJ9Method*) callTarget->_calleeMethod)->getExistingJittedBodyInfo();
+         if (bodyInfo && comp()->getMethodHotness() <= warm && bodyInfo->getHotness() >= hot)
+            continue;
+         }
+#endif
+
+      // The actual symbol for the callTarget->_calleeMethod.
+      TR::ResolvedMethodSymbol* calleeMethodSymbol = TR::ResolvedMethodSymbol::create(comp()->trHeapMemory(), callTarget->_calleeMethod, comp());
+
+      // generate the CFG of this call target and set the block frequencies.
+      TR::CFG* cfg = self()->generateControlFlowGraph(callTarget);
+
+      if (!cfg)
+         {
+         if (traceBIIDTGen)
+            traceMsg(comp(), "Fail to generate a CFG. Don't add\n");
+         continue;
+         }
+
+      if (parent->getRootCallRatio() * callRatio * callTarget->_frequencyAdjustment < COLD_ROOT_CALL_RATIO)
+         continue;
+
+      TR::IDTNode* child = parent->addChild(
+                              _idt->getNextGlobalIDTNodeIndex(),
+                              callTarget,
+                              calleeMethodSymbol,
+                              callSite->_byteCodeIndex,
+                              callRatio * callTarget->_frequencyAdjustment,
+                              _idt->getRegion()
+                              );
+
+      _idt->increaseGlobalIDTNodeIndex();
+      _idt->addCost(child->getCost());
+
+      if (!comp()->incInlineDepth(calleeMethodSymbol, callSite->_bcInfo, callSite->_cpIndex, NULL, !callSite->isIndirectCall(), 0))
+         continue;
+
+      // Build the IDT recursively
+      buildIDT2(child, arguments, callerIndex + 1, child->getBudget(), callStack);
```

Comment:
This `callerIndex` will be used in our OpenJ9 code during interpreting the bytecode when `performAbstractInterpretation()` is called. When the bytecode interpreter need to get the callsite information, it will create a new `TR_ByteCodeInfo` type variable and encapulate it in the callsite information. This `callerIndex` is used to initialize the newly created `TR_ByteCodeInfo` type variable.

Our OpenJ9 code is also submitted for pull request, but it need to be merged after the OMR part being accepted. You may also check our OpenJ9 PR at there https://github.com/eclipse-openj9/openj9/pull/17813.

Reply:
OK, I see what's happening. Because we pass `removeInlinedCallSitesEntry=true` to `decInlineDepth()`, it will remove the inlined call site that was created by `incInlineDepth()`, and a later `incInlineDepth()` can reuse the same inlined call site index. This means that as long as there hasn't been any inlining before the current pass, at this point the inlined site index will be effectively just the inline call stack depth anyway. I think implicitly relying on this property hurts the clarity of the code though

We could pass `comp()->getCurrentInlinedSiteIndex()`, but at that point I think I'd suggest removing the `callerIndex` parameter entirely from `buildIDT2()`, `performAbstractInterpretation()`, `visitCallSite()`, and `addNodesToIDT()`. It will always just be the current inlined site index anyway, and I don't see any use for passing a different value

Reply:
Removed all `callerIndex` paramaters and removed `interpreter.setCallerIndex(callerIndex);` in `performAbstractInterpretation()`. Now, the callerIndex will be automatically set from `comp` when initializing `TR::AbsInterpreter` object.